### PR TITLE
8286451: C2: assert(nb == 1) failed: only when the head is not shared

### PR DESCRIPTION
**Pull request by rwestrel**
nb counts the number of loops that share a single head. The assert
that fires is in code that handles the case of a self loop (a loop
composed of a single block). There can be a self loop and multiple
loops that share a head: the assert makes little sense and I propose
to simply remove it.

I think there's another issue with this code: in the case of a self
loop and multiple loops that share a head, the self loop can be any of
the loop for which the head is cloned not only the one that's passed
as argument to ciTypeFlow::clone_loop_head(). As a consequence, I
moved the logic for self loops in the loop that's applied to all loops
that share the loop head.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286451](https://bugs.openjdk.org/browse/JDK-8286451): C2: assert(nb == 1) failed: only when the head is not shared

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [7ece2dc2](https://git.openjdk.java.net/jdk/pull/8947/files/7ece2dc2eeb9140d7cd1184f888f11482af17a9a)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8947/head:pull/8947` \
`$ git checkout pull/8947`

Update a local copy of the PR: \
`$ git checkout pull/8947` \
`$ git pull https://git.openjdk.java.net/jdk pull/8947/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8947`

View PR using the GUI difftool: \
`$ git pr show -t 8947`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8947.diff">https://git.openjdk.java.net/jdk/pull/8947.diff</a>

</details>

### CODE REVIEW

**Comment by rwestrel** [comment]
> Looks good!

Thanks for the review.

> > nb counts the number of loops that share a single head
> 
> Maybe you also want to rename the variable to make it more clear what it counts.

Done.

File: test/hotspot/jtreg/compiler/ciTypeFlow/TestSharedLoopHead.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8286451
+ * @summary C2: assert(nb == 1) failed: only when the head is not shared
+ * @run main/othervm TestSharedLoopHead
```
[the remark this answers is not in the document]
Comment (rwestrel) [reply test/hotspot/jtreg/compiler/ciTypeFlow/TestSharedLoopHead.java]: Right but the test needs -XX:-BackgroundCompilation

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by rwestrel** [comment]
@chhagedorn @TobiHartmann thanks for the reviews

**Pull request closed without merging**